### ISSUE TITLE
delaying a running job

### ISSUE THREAD

**Issue by blankenshipz** [taskforcesh/bullmq#643]
I have a simple processor like this:

```typescript
    this.worker = new Worker(this.queue.name, async (job: Job) => {
      this.logger.info("Processing job " + job.id)

      const results = await attempt.run(job);

      this.logger.info("Finishing job " + job.id)
      return results
    }
  }
```

Is it possible to call `#Job.moveToDelayed` inside of the processor or will returning from the processor always move it to complete with the return value as the results? 

I'm logically looking for something like this I'd rather not handle the error and completion steps manually if I can avoid it: 

```typescript
    this.worker = new Worker(this.queue.name, async (job: Job) => {
      this.logger.info("Processing job " + job.id)

      const results = await attempt.run(job);

      this.logger.info("Finishing job " + job.id)
      
      if(results.delay) {
        this.logger.info("Delaying job")
        
        # Delay the Job! 
        job.moveToDelayed(results.delay.time)
      }
      
      return results
    }
  }
```

**Comment by manast** [comment]
The job will always be moved to completed/failed. We are exploring ideas on how to be able to express status changes in the processor but we do not have any definitive design yet...

**Comment by blankenshipz** [comment]
Thanks! I'll just write out something to manually process the job for now.

**Comment by blankenshipz** [comment]
@manast Sorry one last question about this; if I handle processing myself do I also have to emit the events? I noticed that in the worker class when it calls `moveToFailed` and `moveToCompleted` it also calls `this.emit`

https://github.com/taskforcesh/bullmq/blob/master/src/classes/worker.ts#L344-L345
https://github.com/taskforcesh/bullmq/blob/master/src/classes/worker.ts#L333-L338

This isn't described as being necessary in these docs: 

https://docs.bullmq.io/patterns/manually-fetching-jobs

**Comment by manast** [comment]
@blankenshipz those are local events. Events using the QueueEvents class will emit even when processing manually.

**Comment by nodkz** [comment]
Before bullmq@1.19.0 we were using the following code without problems (if some conditions are met we delay the job for some period of time): 

```ts
const worker = new Worker(
  'my_queue',
  async (job) => {
    ...
    if (needToDelayConditions) {
      await job.moveToDelayed(Date.now() + 300000);
      return { result: 'delayed' };
    }
    return { result: 'ok' };
  }
);
```

But after bullmq@1.20.2 we met with the following problem:
- `Scripts.moveToDelayed` [wants active job](https://github.com/taskforcesh/bullmq/blob/master/src/commands/moveToDelayed-5.lua#L33-L35) for moving it to delayed list
- After that no matter what we are returning some result or error, in any case, will be called `moveToFinished` method
- Scripts.moveToFinished() also [wants active job](https://github.com/taskforcesh/bullmq/blob/master/src/commands/moveToFinished-8.lua#L65-L67) for moving it completed list
  - but this Job already was removed from the active list by `moveToDelayed`
  - at the end, we receive `-2: "Missing lock for job...".` error
  
  
### How we understood what is happening under the hood now:
  
1. ` job.moveToDelayed` release `:lock` and move a job from `active` to `delay` list.
2. After execution `processFn` the worker calls lua-script `moveToFinished` and tries to move the job from the `active` list (where it no longer exists) to the `completed` list.
3. lua-script `moveToFinished` does not find a job in the `active` list and throws error `-3: "Job is not in the active list ..."`.
4. Worker catches this error,  and again call lua-script `moveToFinished` and in this time tries to move the job to `failed` list
5. lua-script `moveToFinished` does not find `:lock` key (it was removed by the previous `moveToFinished` call) and throws error `-2: "Missing lock for job..."`.

## Expected behavior

`job.moveToDelayed()` method should halt execution of `processFn` without calling `moveToFinished` function. Job is delayed and not finished yet.

**Comment by manast** [comment]
Thanks for reporting. I think the issue is that ```job.moveToDelayed``` should be a private method. The reason is that this method is actually designed for working inside ```job.moveToFinish``` when we need to perform retries.```job.moveToDelayed``` does by the way not call ```job.moveToFinished```.

So actually after 1.19 we fixed the issue of not checking if the job really was on the active state, since that is what is expected when ```job.moveToFinish``` is called and retries are to be performed.

So I am suspecting you are using ```moveToDelayed``` for a different purpose, in that case I would like to know which purpose is that and see if we can find a better solution.

**Comment by nodkz** [comment]
@manast our case is very simple: 3rd party service get requested for metrics every hour, but sometimes metrics are delayed.
So want to have the ability to delay the job which calculates for us the resource consumption for 5 min and try it again later.

Right now (if metrics are not ready) we are creating a new delayed job, finishing the current job with `return { status: 'delayed', new_job_id: ... }`. But it wastes logs of completed jobs. Some jobs are delayed and others are finished with some calculations.

So it will be nice to have the ability to delay job inside processFn, something like:
```ts
const worker = new Worker(
  'my_queue',
  async (job) => {
    // ... obtain source data
    if (sourceDataIsNotReadyYet) {
      return job.delay(Date.now() + 300000);  // returns some `Delay` object which will be handled in special manner by [worker.run](https://github.com/taskforcesh/bullmq/blob/master/src/classes/worker.ts#L139)
    }
    
    // ... calculations ...
    return { result: 'ok' };
  }
);
```

**Comment by manast** [comment]
@nodkz wouldn't we possible to just add a new delayed job and then just return for the current one?

**Comment by nodkz** [comment]
@manast yeah, returning a new delayed job as a result of a current job will make code more obvious. And it will be nice in such case to move the current job into some specific list eg `postponed`. Because moving the current job to `completed` or `failed` lists will look strange.

**Comment by roggervalf** [comment]
this issue is pretty old but I think it should be solved with https://docs.bullmq.io/patterns/process-step-jobs#delaying
